### PR TITLE
A22: Assignments

### PR DESCRIPTION
**Pull request by asfar-nazar**

### CODE REVIEW

File: PSICover/PSICover.cs [hunk 1]
```diff
@@ -146,8 +147,79 @@ class Analyzer {
...
+               <td></td>
+               </tr>
+            """;
+         var files = mBlocks.Where (a => a.Module == module).Select (a => a.File).Distinct ().ToArray ();
+         List<(string file, int tBlocks, int hBlocks, double coverage)> fTable = new ();
+         foreach (var file in files) {
+            var fBlocks = mBlocks.Where (a => a.File == file);
+            int hBlocks = fBlocks.Where (a => hits[a.Id] > 0).Count ();
+            fTable.Add ((file, fBlocks.Count (), hBlocks, GetFileCoverage (file)));
+         }
```
Comment (sdinesh-90) [inline PSICover/PSICover.cs]: `fBlocks.Count()` will go through the mBlocks list to get the total count. Similarly, in `GetFileCoverage (file)`, you are going through the mBlocks list again to determine the percent. You could just pass `fBlocks.Count ()` and hBlocks to the function. Better do the percent calculation here itself.

File: PSICover/PSICover.cs [hunk 2]
```diff
@@ -165,9 +237,16 @@ class Analyzer {
             code[i] = code[i].Replace ('<', '\u00ab').Replace ('>', '\u00bb');
          foreach (var block in blocks) {
             bool hit = hits[block.Id] > 0;
-            string tag = $"<span class=\"{(hit ? "hit" : "unhit")}\">";
-            code[block.ELine] = code[block.ELine].Insert (block.ECol, "</span>");
-            code[block.SLine] = code[block.SLine].Insert (block.SCol, tag);
+            string tag = $"<span class=\"{(hit ? "hit" : "unhit")}\" title = \"Hits: {hits[block.Id]}\">";
+            if (block.ELine > block.SLine) {
```
Comment (sdinesh-90) [inline PSICover/PSICover.cs]: You can achieve the result without the if condition

File: PSICover/PSICover.cs [hunk 3]
```diff
@@ -165,9 +237,16 @@ class Analyzer {
...
          foreach (var block in blocks) {
             bool hit = hits[block.Id] > 0;
-            string tag = $"<span class=\"{(hit ? "hit" : "unhit")}\">";
-            code[block.ELine] = code[block.ELine].Insert (block.ECol, "</span>");
-            code[block.SLine] = code[block.SLine].Insert (block.SCol, tag);
+            string tag = $"<span class=\"{(hit ? "hit" : "unhit")}\" title = \"Hits: {hits[block.Id]}\">";
+            if (block.ELine > block.SLine) {
+               for (int i = block.ELine; i >= block.SLine; i--) {
+                  code[i] = code[i].Insert (code[i].Length, "</span>");
+                  code[i] = code[i].Insert (code[i].TakeWhile (a => a is ' ' or '\t').Count (), tag);
```
Comment (sdinesh-90) [inline PSICover/PSICover.cs]: You could use char.IsWhiteSpace

**Review by sdinesh-90 — CHANGES_REQUESTED** [review]

File: PSICover/PSICover.cs (hunk 3 above)
Comment (asfar-nazar) [reply PSICover/PSICover.cs]: Changed to char.IsWhiteSpace

File: PSICover/PSICover.cs (hunk 2 above)
Comment (asfar-nazar) [reply PSICover/PSICover.cs]: Changed

File: PSICover/PSICover.cs (hunk 1 above)
Comment (asfar-nazar) [reply PSICover/PSICover.cs]: Changed helper method

File: PSICover/PSICover.cs [hunk 4]
```diff
@@ -146,8 +147,80 @@ class Analyzer {
       ExecProgram ($"{Dir}/{RunExe}", "");
    }
 
+   double GetFileCoverage (string file, int blockCount, int hitBlocks) {
+      var blocks = mBlocks.Where (a => a.File == file);
+      ulong[] hits = File.ReadAllLines ($"{Dir}/hits.txt").Select (ulong.Parse).ToArray ();
```
Comment (sdinesh-90) [inline PSICover/PSICover.cs]: blocks and hits are not used. file argument not required.

File: PSICover/PSICover.cs [hunk 5]
```diff
@@ -146,8 +147,80 @@ class Analyzer {
...
+               <th>Module</th>
+               <th>File</th>
+               <th>Total blocks</th>
+               <th>Hit blocks</th>
+               <th>Coverage</th>
+               </tr>
+            """;
+      var modules = mBlocks.Select (a => a.Module).Distinct ().ToArray ();
+      foreach (var module in modules) {
+         text += $$"""
```
Comment (sdinesh-90) [inline PSICover/PSICover.cs]: Use StringBuilder or string.Join as string is readonly and we are creating more temporary string.

File: PSICover/PSICover.cs [hunk 6]
```diff
@@ -165,9 +238,13 @@ class Analyzer {
             code[i] = code[i].Replace ('<', '\u00ab').Replace ('>', '\u00bb');
          foreach (var block in blocks) {
             bool hit = hits[block.Id] > 0;
-            string tag = $"<span class=\"{(hit ? "hit" : "unhit")}\">";
-            code[block.ELine] = code[block.ELine].Insert (block.ECol, "</span>");
-            code[block.SLine] = code[block.SLine].Insert (block.SCol, tag);
+            string tag = $"<span class=\"{(hit ? "hit" : "unhit")}\" title = \"Hits: {hits[block.Id]}\">";
+            for (int i = block.ELine; i >= block.SLine; i--) {
+               int mInsert = (i == block.ELine) ? mInsert = block.ECol : code[i].Length;
```
Comment (sdinesh-90) [inline PSICover/PSICover.cs]: mInsert = is not required before block.ECol. Variable name should not start with mI as it is not class member.

**Review by sdinesh-90 — CHANGES_REQUESTED** [review]

File: PSICover/PSICover.cs (hunk 5 above)
Comment (asfar-nazar) [reply PSICover/PSICover.cs]: Using StringBuilder now

File: PSICover/PSICover.cs (hunk 4 above)
Comment (asfar-nazar) [reply PSICover/PSICover.cs]: Corrected

File: PSICover/PSICover.cs (hunk 6 above)
Comment (asfar-nazar) [reply PSICover/PSICover.cs]: Corrected

**Review by sdinesh-90 — APPROVED** [review]

**Comment by sdinesh-90** [comment]
Looks fine.

**Pull request closed without merging**